### ISSUE TITLE
An alternative way to implement IME for desktop (#500)

### ISSUE THREAD

**Issue by dkwingsmt** [superlistapp/super_editor#591]
We noticed that super_editor recently landed #500 that uses an unusual way to handle keyboard events. Unfortunately, this method will not work properly even for now, and is against the direction Flutter is heading. I will propose a different method.

Your goal, to my understanding, is to make your text widget handle key events only if nobody else in the app handles it. In technical terms, you want to insert another class that handles the key events between the "key manager" and "text input plugin" components of the Flutter engine but on the framework side and aware of the focus system.

However, the current implementation has the following flaws:
- It will not work.
  - `ShortcutManager.shortcuts` is not recursively updated. Querying the closest enclosing `ShortcutManager` will not get the shortcuts registered by outer widgets.
  - It fails to consider non-shortcut key handlers, such as `KeyboardListener`.
- It uses a protected method, `ShortcutManager.handleKeyPress`.
- Flutter is removing `of` and `maybeOf`. This removes a resource-heavy `InheritedWidget` that is rarely used.
  - Actually, this change has been landed as of master (https://github.com/flutter/flutter/pull/104215). We would have reverted it if not for the previous two points.

After discussion, we agreed that your usage is legit albeit rare. We kindly ask you to implement it in a different way: to override `KeyEventManager.keyMessageHandler`. This is the global single entrance to handle key events. Currently it's set by `FocusManager` on binding initialization. However, you can read this handler, and assign it with a new one that executes the current one then yours.

This overriding is an intended design for advanced usage. Overrides must not be undone (since you don't know if others are overriding it too). You must implement your own manager to handle subscribing and unsubscribing. (We'll make these rules clearer in the documentation if you agree with this design.)

To summarize, the proposed way is:
- Make a global manager of your own.
- Override `KeyEventManager`'s callback  on start up by appending this manager's handler to the existing one.
- Make your text fields register and unregister themselves on focus change.
- Call the registered widget for unhandled events.

cc @gspencergoog

**Comment by matthew-carroll** [comment]
> We noticed that super_editor recently landed https://github.com/superlistapp/super_editor/pull/500 that uses an unusual way to handle keyboard events.

We've been working with keyboard input from the beginning. Are you referring to something specific in #500? Which aspects of this request are specific to IME-based input vs our long-standing keyboard input for text entry?

> Unfortunately, this method will not work properly even for now, and is against the direction Flutter is heading

What direction is that? I've been talking to various members of the Flutter team for over a year about this project and I haven't heard anything about a "direction" for text input. Do you have a design doc somewhere?

> Your goal, to my understanding, is to make your text widget handle key events only if nobody else in the app handles it. In technical terms, you want to insert another class that handles the key events between the "key manager" and "text input plugin" components of the Flutter engine but on the framework side and aware of the focus system.

I don't know if we're on the same page here or not. You've described something that is both very abstract, and then very specific. What we're doing is pretty straightforward. We're implementing the ability to edit documents and type into text fields. So that's what we want to accomplish. Which aspects of what you're looking at corresponds to "handle key events only if nobody else in the app handles it"?

> It uses a protected method, ShortcutManager.handleKeyPress

Yeah, I raised this issue a long time ago with Justin, and there should be an issue filed somewhere for this. We were handling keyboard input before that shortcut behavior was introduced. That use of a protected method was a hack to keep Flutter from breaking us. And it's a good thing we were able to use that hack, because it seems that Flutter still hasn't corrected that behavior.

> ShortcutManager.shortcuts is not recursively updated
> Flutter is removing of and maybeOf.

Too bad nobody reached out to the community when those decisions were made...

> After discussion, we agreed that your usage is legit

I would strongly encourage everyone on the team who "discussed" what's "legitimate" to deeply consider the nature of your roles in the Flutter project. The implications of this statement are remarkable, to say the least.

> We kindly ask you to implement it in a different way...

I'm not familiar with the relationships that you're describing. Do you have a minimal example of this special override behavior that you'd like for us to use?

**Comment by dkwingsmt** [comment]
> We've been working with keyboard input from the beginning. Are you referring to something specific in #500? Which aspects of this request are specific to IME-based input vs our long-standing keyboard input for text entry?

Yes, I'm talking about the two use cases of `Shortcuts.maybeOf`. 

> What direction is that? I've been talking to various members of the Flutter team for over a year about this project and I haven't heard anything about a "direction" for text input. Do you have a design doc somewhere?

The direction is to minimize usage of `InheritedWidget`. 

> I don't know if we're on the same page here or not. You've described something that is both very abstract, and then very specific. What we're doing is pretty straightforward. We're implementing the ability to edit documents and type into text fields. So that's what we want to accomplish. Which aspects of what you're looking at corresponds to "handle key events only if nobody else in the app handles it"?

The aspect I'm talking about is the reason why you would like to use `Shortcuts.maybeOf` instead of the more common APIs. In your own words, "trigger any Flutter shortcut for the key combo, and if it does, the editor won't try to handle that key". 

The issue with this approach is that we're not supposed to call the method on ShortcutsManager that runs the shortcut. But I don't see any other way to know if a shortcut is registered.

> Yeah, I raised this issue a long time ago with Justin, and there should be an issue filed somewhere for this. We were handling keyboard input before that shortcut behavior was introduced. That use of a protected method was a hack to keep Flutter from breaking us. And it's a good thing we were able to use that hack, because it seems that Flutter still hasn't corrected that behavior.

Then maybe this proposal can fix that issue in a more intended way.

> Too bad nobody reached out to the community when those decisions were made...

Decisions are made by Flutter's breaking policy. Although I understand that in #500 you mentioned submitting to flutter/tests as soon as possible, the first usage of this protected function in [document_input_keyboard.dart](https://github.com/superlistapp/super_editor/pull/357/files#diff-a9a655d6c2bc7d238d070439b5c388419ee55ed025f4f4e5b152a44dbb89dc14) was merged as long as 6 months ago.

> I would strongly encourage everyone on the team who "discussed" what's "legitimate" to deeply consider the nature of your roles in the Flutter project. The implications of this statement are remarkable, to say the least.

There are certain functionalities that are more proper implemented by 3p libraries. 

> I'm not familiar with the relationships that you're describing. Do you have a minimal example of this special override behavior that you'd like for us to use?

Yes, this overriding behavior is to replace your usage of `Shortcuts.maybeOf`.

**Comment by matthew-carroll** [comment]
@dkwingsmt Do you have a minimal example of the override behavior that you're asking us to use? You said that there's an API specifically designed for this purpose within Flutter. Does Flutter have an example that shows us how to use that override?

**Comment by dkwingsmt** [comment]
I've created https://github.com/flutter/flutter/pull/105280, which includes an example app and refined documentation. The example should be similar to your usage as far as I understand. Let me know if the app and the docs make sense.
